rust/worker/src/blockstore/arrow_blockfile/block/types.rs

```diff
@@ -222,3 +222,324 @@ impl BlockData {
         total_size
     }
 }
+
+// ============== BlockDataBuilder ==============
+
+enum KeyBuilder {
+    StringBuilder(StringBuilder),
+    FloatBuilder(Float32Builder),
+    BoolBuilder(BooleanBuilder),
+}
+
+enum ValueBuilder {
+    Int32ArrayValueBuilder(ListBuilder<Int32Builder>),
+    StringValueBuilder(StringBuilder),
+}
+
+/// BlockDataBuilder is used to build a block. It is used to add data to a block and then build the BlockData once all data has been added.
+/// It is only used internally to an arrow_blockfile.
+pub(super) struct BlockDataBuilder {
+    prefix_builder: StringBuilder,
+    key_builder: KeyBuilder,
+    value_builder: ValueBuilder,
+    last_key: Option<BlockfileKey>,
+}
+
+/// ## Options for the BlockDataBuilder
+/// - item_count: The number of items in the block
+/// - prefix_data_capacity: The required capacity for the prefix data. This will be rounded to the nearest 64 byte alignment by arrow.
+/// - key_data_capacity: The required capacity for the key data. This will be rounded to the nearest 64 byte alignment by arrow.
+/// - total_value_count: The total number of values in the block
```

Comment:
Does the BlockBuilder reject add()s if any of these values would be exceeded? If so could we document that

Reply:
It doesn't, this capacity is preallocated space and will resize. Right now, its the responsibility of the blockfile to manage the size of the blocks. Blocks are size-dumb and can just be allocated with initial space. I guess you could invert this responsibility but I don't have a strong opinion